codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -277,18 +289,17 @@ private boolean isStreamMapEmpty() {
     }
 
     /**
-     * Closed streams may stay in the priority tree if they have dependents that are in prioritizable states.
-     * When a stream is requested to be removed we can only actually remove that stream when there are no more
-     * prioritizable children.
-     * (see [1] {@link Http2Stream#prioritizableForTree()} and [2] {@link DefaultStream#removeChild(DefaultStream)}).
-     * When a priority tree edge changes we also have to re-evaluate viable nodes
-     * (see [3] {@link DefaultStream#takeChild(DefaultStream, boolean, List)}).
-     * @param stream The stream to remove.
+     * Remove a stream from the {@link #streamMap}.
+     * @param stream the stream to remove.
+     * @param itr an iterator that may be pointing to the stream during iteration and {@link Iterator#remove()} will be
+     * used if non-{@code null}.
      */
     void removeStream(DefaultStream stream, Iterator<?> itr) {
-        // [1] Check if this stream can be removed because it has no prioritizable descendants.
         if (stream.parent().removeChild(stream)) {
             if (itr == null) {
+                if (isClosing) {
+                    logger.error("We are removing from the stream map while iterating!", new Throwable());
```

Comment:
@Scottmitch - Sorry for taking so long to response. I've been trying all day long to find a good way to reproduce this problem but still haven't got any progress yet. And for the second question, I think we don't accessing codec-http2 APIs from outside of the eventloop. All the client code which interacting with codec-http2's APIs is [this class](https://github.com/relayrides/pushy/blob/master/src/main/java/com/relayrides/pushy/apns/ApnsClientHandler.java). 

But I do have got something new other than reproducing the NPE. I change [`streamMap` in `DefaultHttp2Connection`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L69) from `IntObjectHashMap` to `java.util.HashMap` (the declaration is `final Map<Integer, Http2Stream> streamMap = new HashMap<Integer, Http2Stream>()` ) I find that the leak and NPE is disappeared. Or at least for four hours I haven't seen it comparing to using `IntObjectHashMap` that NPE always jump out less than an hour in daytime under average load ( I'm at GMT+8 ). But I still need some time to make sure that the NPE is really gone. 

Now I'm trying to do some test for `IntObjectHashMap` to see weather this defect is in `IntObjectHashMap`.